library/Translate/src/Juvix/ToCore/FromFrontend.hs

```diff
@@ -0,0 +1,799 @@
+    args <- traverse isVarArg xs
+    cbody <- transformTermHR p cbody
+    rhs <- toElim (Sexp.Cons (Sexp.atom ":let-match=") e) $ foldr HR.Lam cbody args
+    HR.Let Usage.Omega atomName rhs <$> transformTermHR p body
+transformSimpleLet _p (Sexp.List [_name, fun, _body]) =
+  throwFF $ ExprUnimplemented fun
+transformSimpleLet _ _ = error "malformed let"
+
+isVarArg ::
+  HasThrowFF primTy primVal m =>
+  Sexp.T ->
+  m NameSymbol.T
+isVarArg p@(name Sexp.:> _rest)
+  | Sexp.isAtomNamed name ":implicit-a" =
+    throwFF $ PatternUnimplemented p
+isVarArg p =
+  isVarPat p
+isVarArg _ = error "malformed arg"
+
+isVarPat ::
+  HasThrowFF primTy primVal m =>
+  Sexp.T ->
+  m NameSymbol.T
+isVarPat p
+  | Just Sexp.A {atomName} <- Sexp.atomFromT p =
+    pure atomName
+isVarPat p =
+  throwFF $ PatternUnimplemented p
+
+toElim ::
+  HasThrowFF primTy primVal m =>
+  -- | the original expression
+  Sexp.T ->
+  HR.Term primTy primVal ->
+  m (HR.Elim primTy primVal)
+toElim _ (HR.Elim e) = pure e
+toElim e _ = throwFF $ NotAnElim e
+
+getValSig ::
+  ( HasCoreSigs primTy primVal m,
+    HasThrowFF primTy primVal m
+  ) =>
+  NameSymbol.Mod ->
+  NameSymbol.T ->
+  m (IR.GlobalUsage, HR.Term primTy primVal)
+getValSig q = getSig' q \case ValSig π ty -> Just (π, ty); _ -> Nothing
+
+getConSig ::
+  ( HasCoreSigs primTy primVal m,
+    HasThrowFF primTy primVal m
+  ) =>
+  NameSymbol.Mod ->
+  NameSymbol.T ->
+  m (HR.Term primTy primVal, Maybe (Ctx.Def Sexp.T Sexp.T))
+getConSig q = getSig' q \case
+  ConSig (Just ty) def -> Just (ty, def)
+  _ -> Nothing
+
+getDataSig ::
+  ( HasCoreSigs primTy primVal m,
+    HasThrowFF primTy primVal m
+  ) =>
+  NameSymbol.Mod ->
+  NameSymbol.T ->
+  m (HR.Term primTy primVal, [NameSymbol.T])
+getDataSig q = getSig' q \case DataSig ty cons -> Just (ty, cons); _ -> Nothing
+
+getSig' ::
+  ( HasCoreSigs primTy primVal m,
+    HasThrowFF primTy primVal m
+  ) =>
+  NameSymbol.Mod ->
+  (CoreSigHR primTy primVal -> Maybe a) ->
+  NameSymbol.T ->
+  m a
+getSig' q f x = do
+  msig <- lookupSig (Just q) x
+  case msig of
+    Just sig | Just ty <- f sig -> pure ty
+    _ -> throwFF $ WrongSigType x msig
+
+lookupSig ::
+  HasCoreSigs primTy primVal m =>
+  Maybe NameSymbol.Mod -> -- namespace of current declaration
+  NameSymbol.T ->
+  m (Maybe (CoreSig' HR.T primTy primVal))
+lookupSig q x = fmap snd <$> lookupSig' q x
+
+conDefName :: NameSymbol.T -> NameSymbol.T
+conDefName = NameSymbol.applyBase (<> "$def")
+
+transformType ::
+  ( HasPatVars m,
+    HasNextPatVar m,
+    ReduceEff primTy primVal m
+  ) =>
+  NameSymbol.Mod ->
+  NameSymbol.T ->
+  a ->
+  m [IR.RawGlobal primTy primVal]
+transformType q name _ = do
+  (ty, conNames) <- getDataSig q name
+  let getConSig' x = do (ty, def) <- getConSig q x; pure (x, ty, def)
+  conSigs <- traverse getConSig' conNames
+  cons <- traverse (uncurry3 $ transformCon q) conSigs
+  (args, ℓ) <- splitDataType name ty
+  let dat' =
+        IR.RawDatatype
```

Comment:
The field `rawDataPos` is missing here (which was already true, sorry, but might as well fix it now).

Until the frontend has an idea of parameters vs indices, i suppose it has to just be `[]` as a placeholder though